src/hotspot/share/runtime/mutexLocker.cpp

```diff
@@ -169,9 +169,6 @@ void assert_locked_or_safepoint(const Mutex* lock) {
   if (lock->owned_by_self()) return;
   if (SafepointSynchronize::is_at_safepoint()) return;
   if (!Universe::is_fully_initialized()) return;
```

Comment:
Is this bailout on L171 for the code you are deleting (old L172-4)
or for the fatal() on old L175?

Reply:
I was hope that "git blame" would help, but most of this code is ancient:

da18495f385b src/hotspot/share/runtime/mutexLocker.cpp    (Coleen Phillimore       2019-08-22 09:51:36 -0400 166) void assert_locked_or_safepoint(const Mutex* lock) {
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 167)   // check if this thread owns the lock (common case)
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 168)   assert(lock != NULL, "Need non-NULL lock");
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 169)   if (lock->owned_by_self()) return;
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 170)   if (SafepointSynchronize::is_at_safepoint()) return;
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 171)   if (!Universe::is_fully_initialized()) return;
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 172)   // see if invoker of VM operation owns it
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 173)   VM_Operation* op = VMThread::vm_operation();
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 174)   if (op != NULL && op->calling_thread() == lock->owner()) return;
1e71f6773659 hotspot/src/share/vm/runtime/mutexLocker.cpp (David Lindholm          2015-09-29 11:02:08 +0200 175)   fatal("must own lock %s", lock->name());
8153779ad32d hotspot/src/share/vm/runtime/mutexLocker.cpp (J. Duke                 2007-12-01 00:00:00 +0000 176) }

Reply:
The initialization bail-out was added "1.86 99/02/19 17:05:07", but no comments or associated bug report.

Reply:
Universe is initialized before VM thread is ready to process operations.
So while that bail-out is true, there can never be any vm ops.

I believe instead of:
`if (!Universe::is_fully_initialized()) return;`
We really want:
`if (Threads::number_of_threads() == 0) return;`
But I'm not sure and totally out of scope.